src/NuGet.Core/NuGet.Credentials/DelegatingLogger.cs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+
+namespace NuGet.Credentials
+{
+    /// <summary>
+    ///  A delegating logger.
+    /// </summary>
+    internal class DelegatingLogger : LoggerBase, ILogger
+    {
+        private readonly SemaphoreSlim _semaphore;
+        private ILogger _delegateLogger;
+
+        internal DelegatingLogger(ILogger delegateLogger) : base()
+        {
+            _delegateLogger = delegateLogger ?? throw new ArgumentNullException(nameof(delegateLogger));
+            _semaphore = new SemaphoreSlim(1, 1);
+        }
+
+        internal void UpdateDelegate(ILogger logger)
+        {
+            try
+            {
+                _semaphore.Wait();
```

Comment:
I think the `Wait`'s should be outside the try block. If it thows for any reason, then the lock was not obtained, but the finally block will still release.

I previously looked at `SemaphoreSlim`'s code, and it seems like the only reason it will throw is you pass a `CancellationToken` which gets cancelled before the wait completes. Since no `CancellationToken` is used here, it might be impossible for it to throw, but I think as a rule the waits should be outside the try block.

Reply:
yeah it's a good pattern. 
Fixing that.